### PR TITLE
feat: add R3 adaptation

### PR DESCRIPTION
**Pull request by WojtAcht**
image

### CODE REVIEW

File: src/helpers/factories.py [hunk 1]
```diff
@@ -26,12 +27,17 @@ def problem_factory(problem: EProblems):
     return problem_classes[problem]()
 
 
-def adaptation_factory(adaptation: EAdaptations, base_points: torch.Tensor, x_range: [float, float],
-                       max_number_of_points: int=params.NUM_MAX_POINTS):
+def adaptation_factory(
```
Comment (WojtAcht) [inline src/helpers/factories.py:30]: Sorry for this formatting changes. Do you use any linter/formatter? Wdyt about using pre-commit? I can set it up.

File: src/params/params.py [hunk 2]
```diff
@@ -32,20 +32,22 @@
...
-
-
-
-
-
-
+if torch.cuda.is_available():
+    DEVICE = torch.device("cuda")
+    print("Using CUDA")
+elif torch.backends.mps.is_available():
```
Comment (WojtAcht) [inline src/params/params.py:44]: I added support for `mps`. It's interesting that the `TAN_03` fails for me:
```
Traceback (most recent call last):
  File "/Users/wojciechachtelik/Documents/studia/doktorat/Adaptive_PINN_1D/main.py", line 19, in <module>
    train_PINN(
  File "/Users/wojciechachtelik/Documents/studia/doktorat/Adaptive_PINN_1D/src/runners/adaptive_PINN.py", line 58, in train_PINN
    stage_convergence_data = train_model(nn_approximator=pinn, loss_fn=loss_fn, device=params.DEVICE, learning_rate=params.LEARNING_RATE,
  File "/Users/wojciechachtelik/Documents/studia/doktorat/Adaptive_PINN_1D/src/base/pinn_core.py", line 90, in train_model
    loss = loss_fn(pinn=nn_approximator)
  File "/Users/wojciechachtelik/Documents/studia/doktorat/Adaptive_PINN_1D/src/problems/tan_03.py", line 28, in compute_loss
    assert (boundary_right == self.range[1]).all().item(), f"Last point not on a boundary: {boundary_right}"
AssertionError: Last point not on a boundary: tensor([[1.5708]], device='mps:0', grad_fn=<ViewBackward0>)
```
Probably the reason is the fact that `mps` doesn't support `torch.float64`.

File: src/helpers/factories.py (hunk 1 above)
Comment (JanTry) [reply src/helpers/factories.py:30]: I've used one before, but for this repo I've rewritten everything and I had no time to do multiple things I wanted 

Feel free to edit whatever you want :D

File: src/params/params.py (hunk 2 above)
Comment (JanTry) [reply src/params/params.py:43]: I haven't experienced the issue 
We can either debug that together or I can assist you with all needed computation with my setup and GPU

**Review by JanTry — APPROVED** [review]
Nice input! Feel free to merge anytime you want 

Also, if you need to merge asap feel free to either reach out to me on discord or just merge without a CR 

Ps. I've worked with you before and I would vouch for your code anytime ;)

**Pull request merged**